File: VirtualOrigenApp/virtual_origen_app/lib/models/smart_device.dart
import 'package:flutter/material.dart';
import 'package:uuid/uuid.dart';
import 'package:virtual_origen_app/models/pair.dart';
import 'package:virtual_origen_app/models/smart_device_type.dart';
import 'package:virtual_origen_app/themes/colors.dart';

class SmartDevice {
  final String id;
  final String name;
  final SmartDeviceType type;
  final MyColors color;
  final List<DateTimeRange> timeZones;
  final List<bool> days;
  final Pair<int, int>? batteryRange;
  final Pair<int, int>? productionRange;
  final Pair<int, int>? consumptionRange;
  final Pair<int, int>? temperatureRange;
  final Pair<int, int>? rainRange;
  final bool isManualMode;
  final bool isOn;

  SmartDevice({
    String? id,
    required this.name,
    required this.type,
    required this.color,
    required this.timeZones,
    required this.days,
    this.batteryRange,
    this.productionRange,
    this.consumptionRange,
    this.temperatureRange,
    this.rainRange,
    this.isManualMode = false,
    this.isOn = false,
  }) : id = id ?? const Uuid().v4();

  SmartDevice.fromJson(Map<String, dynamic> json)
      : id = json["id"],
        name = json["name"],
        type = SmartDeviceType.values
            .firstWhere((element) => element.token == json["type"]),
        color = MyColors.values
            .firstWhere((element) => element.token == json["color"]),
        timeZones = json["timeZones"]
            .map<DateTimeRange>((e) => DateTimeRange(
                  start: DateTime.parse(e.split(" -")[0]),
                  end: DateTime.parse(e.split("- ")[1]),
                ))
            .toList(),
        days = (json['days'] as List<dynamic>).cast<bool>(),
        batteryRange = json.containsKey("batteryRange")
            ? Pair.intFromString(json["batteryRange"])
            : null,
        productionRange = json.containsKey("productionRange")
            ? Pair.intFromString(json["productionRange"])
            : null,
        consumptionRange = json.containsKey("consumptionRange")
            ? Pair.intFromString(json["consumptionRange"])
            : null,
        temperatureRange = json.containsKey("temperatureRange")
            ? Pair.intFromString(json["temperatureRange"])
            : null,
        rainRange = json.containsKey("rainRange")
            ? Pair.intFromString(json["rainRange"])
            : null,
        isManualMode =
            json.containsKey("isManualMode") ? json["isManualMode"] : false,
        isOn = json["isOn"];

  copyWith({
    String? name,
    SmartDeviceType? type,
    MyColors? color,
    List<DateTimeRange>? timeZones,
    List<bool>? days,
    Pair<int, int>? batteryRange,
    Pair<int, int>? productionRange,
    Pair<int, int>? consumptionRange,
    Pair<int, int>? temperatureRange,
    Pair<int, int>? rainRange,
    bool? isManualMode,
    bool? isOn,
  }) {
    return SmartDevice(
      id: id,
      name: name ?? this.name,
      type: type ?? this.type,
      color: color ?? this.color,
      timeZones: timeZones ?? this.timeZones,
      days: days ?? this.days,
      batteryRange: batteryRange ?? this.batteryRange,
      productionRange: productionRange ?? this.productionRange,
      consumptionRange: consumptionRange ?? this.consumptionRange,
      temperatureRange: temperatureRange ?? this.temperatureRange,
      rainRange: rainRange ?? this.rainRange,
      isManualMode: isManualMode ?? this.isManualMode,
      isOn: isOn ?? this.isOn,
    );
  }

  Map<String, dynamic> toJson() {
    var map = {
      "id": id,
      "name": name,
      "type": type.token,
      "color": color.token,
      "timeZones": timeZones.map((e) => e.toString()).toList(),
      "days": days,
      "isManualMode": isManualMode,
      "isOn": isOn,
    };
    if (batteryRange != null) {
      map["batteryRange"] = batteryRange!.toString();
    }
    if (productionRange != null) {
      map["productionRange"] = productionRange!.toString();
    }
    if (consumptionRange != null) {
      map["consumptionRange"] = consumptionRange!.toString();
    }
    if (temperatureRange != null) {
      map["temperatureRange"] = temperatureRange!.toString();
    }
    if (rainRange != null) {
      map["rainRange"] = rainRange!.toString();
    }
    return map;
  }
}

File: VirtualOrigenServer/VirtualOrigenServer/SmartDevicesService/Models/SmartDevice.cs
using Google.Cloud.Firestore;
using System.Text.Json.Serialization;

namespace SmartDevicesService.Models;

[FirestoreData]
public class SmartDevice
{
    [FirestoreProperty(name: "id")]
    [property: JsonPropertyName("id")]
    public string Id { get; private set; }

    [FirestoreProperty(name: "name")]
    [property: JsonPropertyName("name")]
    public string Name { get; private set; }

    [FirestoreProperty(name: "timeZones")]
    [property: JsonPropertyName("timeZones")]
    public List<string> TimeZones { get; private set; }

    [FirestoreProperty(name: "days")]
    [property: JsonPropertyName("days")]
    public List<bool> Days { get; private set; }

    [FirestoreProperty(name: "batteryRange")]
    [property: JsonPropertyName("batteryRange")]
    public string? BatteryRange { get; private set; }

    [FirestoreProperty(name: "productionRange")]
    [property: JsonPropertyName("productionRange")]
    public string? ProductionRange { get; private set; }

    [FirestoreProperty(name: "consumptionRange")]
    [property: JsonPropertyName("consumptionRange")]
    public string? ConsumptionRange { get; private set; }

    [FirestoreProperty(name: "temperatureRange")]
    [property: JsonPropertyName("temperatureRange")]
    public string? TemperatureRange { get; private set; }

    [FirestoreProperty(name: "rainRange")]
    [property: JsonPropertyName("rainRange")]
    public string? RainRange { get; private set; }

    [FirestoreProperty(name: "isManualMode")]
    [property: JsonPropertyName("isManualMode")]
    public bool IsManualMode { get; private set; }

    [FirestoreProperty(name: "isOn")]
    [property: JsonPropertyName("isOn")]
    public bool IsOn { get; private set; }

    [FirestoreProperty(name: "type")]
    [property: JsonPropertyName("type")]
    public string Type { get; private set; }
    [FirestoreProperty(name: "color")]
    [property: JsonPropertyName("color")]
    public string Color { get; private set; }

    [JsonConstructor]
    public SmartDevice(string id, string name, List<string> timeZones, List<bool> days, bool isManualMode, bool isOn, string? batteryRange = null, string? productionRange = null, string? consumptionRange = null, string? temperatureRange = null, string? rainRange = null, string type = "other", string color = "primary")
    {
        Id = id;
        Name = name;
        TimeZones = timeZones;
        Days = days;
        BatteryRange = batteryRange;
        ProductionRange = productionRange;
        ConsumptionRange = consumptionRange;
        TemperatureRange = temperatureRange;
        RainRange = rainRange;
        IsManualMode = isManualMode;
        IsOn = isOn;
        Type = type;
        Color = color;
    }

    public SmartDevice(Dictionary<string, object> data)
    {
        Id = data["id"] as string ?? "";
        Name = data["name"] as string ?? "";
        TimeZones = ((List<object>)data["timeZones"]).Cast<string>().ToList() ?? [];
        Days = ((List<object>)data["days"]).Cast<bool>().ToList() ?? [];
        BatteryRange = data.TryGetValue("batteryRange", out object? value) ? value as string : null;
        ProductionRange = data.TryGetValue("productionRange", out value) ? value as string : null;
        ConsumptionRange = data.TryGetValue("consumptionRange", out value) ? value as string : null;
        TemperatureRange = data.TryGetValue("temperatureRange", out value) ? value as string : null;
        RainRange = data.ContainsKey("rainRange") ? data["rainRange"] as string : null;
        IsManualMode = data["isManualMode"] as bool? ?? false;
        IsOn = data["isOn"] as bool? ?? false;
        Type = data.TryGetValue("type", out value) ? value as string : "other";
        Color = data.TryGetValue("color", out value) ? value as string : "primary";
    }

    public SmartDevice()
    {
        Id = "";
        Name = "";
        TimeZones = [];
        Days = [];
        BatteryRange = null;
        ProductionRange = null;
        ConsumptionRange = null;
        TemperatureRange = null;
        RainRange = null;
        IsManualMode = false;
        IsOn = false;
        Type = "other";
        Color = "primary";
    }

    public void SetIsOn(bool isOn)
    {
        IsOn = isOn;
    }

    public Dictionary<string, object> keyValues()
    {
        var data = new Dictionary<string, object>
        {
            { "id", Id },
            { "name", Name },
            { "timeZones", TimeZones },
            { "days", Days },
            { "isManualMode", IsManualMode },
            { "isOn", IsOn },
            { "type", Type },
            { "color", Color }
        };
        if (BatteryRange != null)
        {
            data.Add("batteryRange", BatteryRange);
        }
        if (ProductionRange != null)
        {
            data.Add("productionRange", ProductionRange);
        }
        if (ConsumptionRange != null)
        {
            data.Add("consumptionRange", ConsumptionRange);
        }
        if (TemperatureRange != null)
        {
            data.Add("temperatureRange", TemperatureRange);
        }
        if (RainRange != null)
        {
            data.Add("rainRange", RainRange);
        }
        return data;
    }
}
